s3stream/src/main/java/com/automq/stream/s3/operator/AbstractObjectStorage.java

```diff
@@ -170,29 +180,71 @@ public CompletableFuture<WriteResult> write(WriteOptions options, String objectP
         return retCf;
     }
 
+    private void recordWriteStats(String path, long objectSize, TimerUtil timerUtil) {
+        s3LatencyCalculator.record(objectSize, timerUtil.elapsedAs(TimeUnit.MILLISECONDS));
+        S3OperationStats.getInstance().uploadSizeTotalStats.add(MetricsLevel.INFO, objectSize);
+        S3OperationStats.getInstance().putObjectStats(objectSize, true).record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("put object {} with size {}, cost {}ms", path, objectSize, timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
+        }
+    }
+
     private void write0(WriteOptions options, String path, ByteBuf data, CompletableFuture<Void> cf) {
         TimerUtil timerUtil = new TimerUtil();
         long objectSize = data.readableBytes();
-        doWrite(options, path, data).thenAccept(nil -> {
-            S3OperationStats.getInstance().uploadSizeTotalStats.add(MetricsLevel.INFO, objectSize);
-            S3OperationStats.getInstance().putObjectStats(objectSize, true).record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug("put object {} with size {}, cost {}ms", path, objectSize, timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
-            }
+
+        CompletableFuture<Void> writeCf = doWrite(options, path, data);
+        AtomicBoolean completedFlag = new AtomicBoolean(false);
+        WriteOptions retryOptions = options.copy().retry(true);
+
+        // Fast retry should only be triggered by the original request.
+        long delayMillis = s3LatencyCalculator.valueAtPercentile(objectSize, 99);
+        if (options.enableFastRetry() && delayMillis > 0 && !options.retry()) {
+            scheduler.schedule(() -> {
```

Comment:
You could use HashedWheelTimer to schedule and cancel the task. HashedWheelTimer has better performance